sklearn/utils/estimator_checks.py

```diff
@@ -385,55 +394,244 @@ def check_transformer_pickle(name, Transformer):
+                          cloned.partial_fit(X, y, classes), m2)
+
+    elif isinstance(estimator, RegressorMixin):
+        ret_val = estimator.partial_fit(X, y)
+        assert_same_model(X, estimator, cloned.partial_fit(X, y), m2)
+
     else:
-        alg.partial_fit(X, y)
+        ret_val = estimator.partial_fit(X)
+        assert_same_model(X, estimator, cloned.partial_fit(X), m2)
 
-    assert_raises(ValueError, alg.partial_fit, X[:, :-1], y)
+    # Check if returns self
+    assert_equal(ret_val, estimator, msg=m1)
 
 
-def check_clustering(name, Alg):
+def check_partial_fit_reset_when_fit(name, Estimator):
+    """Check if the estimators get reset when fit is applied after a 
+    partial fit.
+    """
+   
+    # Estimators are assumed to be reset when the transorm / predic result
+    # is similar to 1st fit call
+    if not hasattr(Estimator, 'partial_fit'):
+        return
+
+    X, y = make_blobs(n_samples=50, random_state=1)
+    X -= X.min()
+
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
+   
+    set_random_state(estimator)
+    set_fast_parameters(estimator)
+    
+    classes = np.unique(y)
+
+    # Initial call to fit
+    if isinstance(estimator, (ClassifierMixin, RegressorMixin)):
+        estimator.fit(X, y)
+    else:
+        estimator.fit(X)
+    
+    # Save the model in current state (after one fit call)
+    model_after_first_fit = deepcopy(estimator)
+
+    # Run 2 iterations of full set of partial_fit with batches of size 10
+    for i in range(2):
+        X_s, y_s = shuffle(X, y, random_state=0)
+        for j in range(0, 50, 10):
+            if isinstance(estimator, ClassifierMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j],
+                                      classes=classes)
+            elif isinstance(estimator, RegressorMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j])
+            else:
+                estimator.partial_fit(X_s[j:10+j])
+   
+    # Make sure that the partial_fit model is not same as first fit model
+    # assert_not_same_model(X, estimator, model_after_first_fit) 
+
+    emsg = "Change in features during 2nd fit raised an error.\n" 
+
+    # Use fit(x, y) if classifier / regressor.
+    if isinstance(estimator, (ClassifierMixin, RegressorMixin)):
+        # Test that change in features does not raise any error
+        try:
+            estimator.fit(X[:, :-1], y)
+        except Exception as e:
+            # Change in features during the second fit call raised an error
+            assert False, emsg + str(e)
+        # Finally call fit to reset the estimator
+        estimator.fit(X, y)
+
+    # Use fit(x) otherwise.
+    else:
+        try:
+            estimator.fit(X[:, :-1])
+        except Exception as e:
+            assert False, emsg + str(e)
+        estimator.fit(X)
+
+    assert_same_model(X, estimator, model_after_first_fit)
+
+
+def check_partial_fit_effect(name, Estimator):
```

Comment:
Sure, you could have asked me there and then! But it's not at all the same as the test you've presented here.

The point there is to ensure that the `partial_fit` does not do the same thing that `fit` does, i.e. overwrite the model. It does so by checking that if I `partial_fit` on some small amount of data *after `partial_fit`ting on a large amount of data* then I will get a different result to just `partial_fit`ting on that small amount of data alone. Realise that it's important that the control model is trained only on a small amount, because if you are trying to contrast `X` and `X[:-1]` (let alone `X` augmented by `X[:-1]`) for n_samples not-so-small, there's a very good chance that the model already has a decent fit.

Finally, in general I wouldn't recommend ever using the training `X` in `assert[_not]_same_model` (and I would document as much in those functions. It produces a very biased idea of "same model" since many models (without parameters tuned to reduce variance) will gladly overfit to their training data.

Your use of deepcopy also adds to the obfuscation.